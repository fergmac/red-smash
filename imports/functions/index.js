// A file for importing some commonly used functions

// a silly little function to add a suffix to the end of a number (e.g. '21' to '21st')
export const numberSuffixer = (num) => {
  if (num.toString().match(/[^\d]/) || num.length>100000000000) {
    return num
  }

  const modulo100 = (num % 100).toString()

  switch (modulo100.substr(0, 1)) {
    case '1':
      if (num > 1 && modulo100.length>1) {
        return num + 'th'
      }
    default:
      switch (num % 10) {
        case 1:
          return num + 'st'
        case 2:
          return num + 'nd'
        case 3:
          return num + 'rd'
        default:
          return num + 'th'
      }
  }
}

export const sortByKey = (key) => (a, b) => {
  switch (true) {
    case a[key] < b[key]:
      return 1;
    case a[key] > b[key]:
      return -1;
    default:
      return 0;
  }
}
