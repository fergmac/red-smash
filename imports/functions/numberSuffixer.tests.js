import { Meteor } from 'meteor/meteor'
import { chai } from 'meteor/practicalmeteor:chai'
import { numberSuffixer } from './'

if (Meteor.isClient) {
  describe('reusable functions - numberSuffixer', function () {
    it('converts the number 721 to \'721st\'', function () {
      chai.assert.strictEqual(numberSuffixer(721), '721st')
    })

    it('converts the string 721 to \'721st\'', function () {
      chai.assert.strictEqual(numberSuffixer('721'), '721st')
    })

    it('converts the number 211 to \'211th\'', function () {
      chai.assert.strictEqual(numberSuffixer(211), '211th')
    })

    it('converts the string 1000000000000001 to \'1000000000000001st\'', function () {
      chai.assert.strictEqual(numberSuffixer('1000000000000001'), '1000000000000001st')
    })

    it('fails to convert the number 10000000000000000000000002 to \'10000000000000000000000002nd\'', function () {
      chai.assert.strictEqual(numberSuffixer(10000000000000000000000002), 10000000000000000000000002)
    })

    it('fails to convert the string 10000000000000000000000002 to \'10000000000000000000000002nd\' because of floating point math error-handling', function () {
      chai.assert.strictEqual(numberSuffixer(10000000000000000000000002), 10000000000000000000000002)
    })

    it('returns the number -211 as -211', function () {
      chai.assert.strictEqual(numberSuffixer(-211), -211)
    })

    it('returns the number 102.4 as 102.4', function () {
      chai.assert.strictEqual(numberSuffixer(102.4), 102.4)
    })

    it('returns the string \'102.4\' as \'102.4\'', function () {
      chai.assert.strictEqual(numberSuffixer('102.4'), '102.4')
    })

    it('returns the string \'face\' as \'face\'', function () {
      chai.assert.strictEqual(numberSuffixer('face'), 'face')
    })

    it('returns an object that is equal to the input (using deepEqual)', function () {
      chai.assert.deepEqual(numberSuffixer({ face: 'feet' }), { face: 'feet' })
    })
  })
}
