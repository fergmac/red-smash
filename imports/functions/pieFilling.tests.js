// import { lookingFor } from './teamStarsFinder.tests'
import { Meteor } from 'meteor/meteor'
import { chai } from 'meteor/practicalmeteor:chai'
import { pieFilling } from './'

const lookingForPie = [
  {
    color: "#9B121E",
    highlight: "#9B121E",
    label: "AD",
    labelColor: "white",
    labelFontSize: "16px",
    value: 129
  },
  {
    color: "#CB3B46",
    highlight: "#CB3B46",
    label: "DM",
    labelColor: "white",
    labelFontSize: "16px",
    value: 70
  },
  {
    color: "#F0AD4E",
    highlight: "#F0AD4E",
    label: "UX",
    labelColor: "white",
    labelFontSize: "16px",
    value: 67
  },
  {
    color: "#F0D64E",
    highlight: "#F0D64E",
    label: "WD",
    labelColor: "white",
    labelFontSize: "16px",
    value: 25
  },
  {
    color: "#222222",
    highlight: "#222222",
    label: "UI",
    labelColor: "white",
    labelFontSize: "16px",
    value: 15
  }
]

const lookingFor = [
  {
    players: [
      {
        _id: "MRz6MwYfErHo9Aamx",
        starCount: 35,
        teamId: "AD",
        username: "Tracey"
      },
      {
        _id: "msE6HxEEw6hCNLJra",
        starCount: 30,
        teamId: "AD",
        username: "Brandon"
      },
      {
        _id: "25EERbFqnQzcFektr",
        starCount: 30,
        teamId: "AD",
        username: "Anna"
      },
      {
        _id: "DwzB2jp5fDcZK26wq",
        starCount: 24,
        teamId: "AD",
        username: "Robert"
      },
      {
        _id: "ta8WSE5ermcQyBxGX",
        starCount: 10,
        teamId: "AD",
        username: "Jody"
      },
    ],
    starCount: 129,
    teamId: "AD"
  },
  {
    players: [
      {
        _id: "k3ezgBDwDZYj6wAZR",
        starCount: 40,
        teamId: "DM",
        username: "Fergus"
      },
      {
        _id: "WuyWdjRJNsAxegAuK",
        starCount: 30,
        teamId: "DM",
        username: "Quinlan"
      },
    ],
    starCount: 70,
    teamId: "DM"
  },
  {
    players: [
      {
        _id: "47YAyGoFTQ7wkztmj",
        starCount: 55,
        teamId: "UX",
        username: "Scott"
      },
      {
        _id: "iyoAjEoW7SApwjAaP",
        starCount: 12,
        teamId: "UX",
        username: "ScottFace"
      },
    ],
    starCount: 67,
    teamId: "UX"
  },
  {
    players: [
      {
        _id: "EGuZWhfieYvReZwCi",
        starCount: 25,
        teamId: "WD",
        username: "Sharifa"
      },
      {
        _id: "BaPqqbzf6uvTdfwAE",
        starCount: 0,
        teamId: "WD",
        username: "Kevin"
      },
    ],
    starCount: 25,
    teamId: "WD"
  },
  {
    players: [
      {
        _id: "d4NdHy3Nxqku7MN7S",
        starCount: 15,
        teamId: "UI",
        username: "Amir"
      },
    ],
    starCount: 15,
    teamId: "UI"
  },
]

if (Meteor.isClient) {
  describe('reusable functions - pieFilling', function () {
    it('should take an array of objects, with a particular layout, as its input, do some work on the array, and compare its output with a manually entered array of objects, checking if they\'re deeply equal', function () {
      const pieFillingOutput = pieFilling(lookingFor)
      chai.assert.deepEqual(pieFillingOutput, lookingForPie)
    })
    
    it('should not be deeply equal to a manually entered array that matches with pieFilling\'s output but also has an additional object pushed onto the end of the array', function () {
      const pieFillingOutput = pieFilling(lookingFor)
      let modifiedLookingForPie = lookingForPie
      modifiedLookingForPie.push({label: 'face', value: 666})
      chai.assert.notDeepEqual(pieFillingOutput, modifiedLookingForPie)
    })
  })
}
