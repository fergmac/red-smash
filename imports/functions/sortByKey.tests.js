import { Meteor } from 'meteor/meteor'
import { chai } from 'meteor/practicalmeteor:chai'
import { sortByKey } from './'
import { numberSuffixer } from './'


if (Meteor.isClient) {
  describe('reusable functions - sortByKey', function () {
    const arrayToSort = [
      {
        teamId: 'AD',
        starCount: 14
      },
      {
        teamId: 'UX',
        starCount: 9
      },
      {
        teamId: 'WD',
        starCount: 17
      },
      {
        teamId: 'UI',
        starCount: 21
      },
      {
        teamId: 'DM',
        starCount: 12
      },
    ]

    const myManualSortLeastToGreatest = [
      {
        teamId: 'UX',
        starCount: 9
      },
      {
        teamId: 'DM',
        starCount: 12
      },
      {
        teamId: 'AD',
        starCount: 14
      },
      {
        teamId: 'WD',
        starCount: 17
      },
      {
        teamId: 'UI',
        starCount: 21
      },
    ]

    const myManualSortGreatestToLeast = [
      {
        teamId: 'UI',
        starCount: 21
      },
      {
        teamId: 'WD',
        starCount: 17
      },
      {
        teamId: 'AD',
        starCount: 14
      },
      {
        teamId: 'DM',
        starCount: 12
      },
      {
        teamId: 'UX',
        starCount: 9
      },
    ]

    it('sorts an example array of objects by the object key, least to greatest', function () {
      chai.assert.deepEqual(arrayToSort.sort(sortByKey('starCount', -1)), myManualSortLeastToGreatest)
    })

    it('sorts an example array of objects by the object key, greatest to least', function () {
      chai.assert.deepEqual(arrayToSort.sort(sortByKey('starCount')), myManualSortGreatestToLeast)
    })
  })
}