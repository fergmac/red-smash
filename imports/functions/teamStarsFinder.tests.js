import { Meteor } from 'meteor/meteor'
import { chai } from 'meteor/practicalmeteor:chai'
import { teamStarsFinder } from './'

const teams = [
  {
    "_id": "ta8WSE5ermcQyBxGX",
    "username": "Jody",
    "teamId": "AD",
    "starCount": 10
  },
  {
    "_id": "d4NdHy3Nxqku7MN7S",
    "username": "Amir",
    "teamId": "UI",
    "starCount": 15
  },
  {
    "_id": "BaPqqbzf6uvTdfwAE",
    "username": "Kevin",
    "teamId": "WD",
    "starCount": 0
  },
  {
    "_id": "DwzB2jp5fDcZK26wq",
    "username": "Robert",
    "teamId": "AD",
    "starCount": 24
  },
  {
    "_id": "MRz6MwYfErHo9Aamx",
    "username": "Tracey",
    "starCount": 35,
    "teamId": "AD"
  },
  {
    "_id": "k3ezgBDwDZYj6wAZR",
    "username": "Fergus",
    "starCount": 40,
    "teamId": "DM"
  },
  {
    "_id": "msE6HxEEw6hCNLJra",
    "username": "Brandon",
    "starCount": 30,
    "teamId": "AD"
  },
  {
    "_id": "25EERbFqnQzcFektr",
    "username": "Anna",
    "starCount": 30,
    "teamId": "AD"
  },
  {
    "_id": "WuyWdjRJNsAxegAuK",
    "username": "Quinlan",
    "starCount": 30,
    "teamId": "DM"
  },
  {
    "_id": "EGuZWhfieYvReZwCi",
    "username": "Sharifa",
    "starCount": 25,
    "teamId": "WD"
  },
  {
    "_id": "47YAyGoFTQ7wkztmj",
    "username": "Scott",
    "starCount": 55,
    "teamId": "UX"
  },
  {
    "_id": "iyoAjEoW7SApwjAaP",
    "username": "ScottFace",
    "starCount": 12,
    "teamId": "UX"
  }]


const lookingFor = [
  {
    players: [
      {
        _id: "MRz6MwYfErHo9Aamx",
        starCount: 35,
        teamId: "AD",
        username: "Tracey"
      },
      {
        _id: "msE6HxEEw6hCNLJra",
        starCount: 30,
        teamId: "AD",
        username: "Brandon"
      },
      {
        _id: "25EERbFqnQzcFektr",
        starCount: 30,
        teamId: "AD",
        username: "Anna"
      },
      {
        _id: "DwzB2jp5fDcZK26wq",
        starCount: 24,
        teamId: "AD",
        username: "Robert"
      },
      {
        _id: "ta8WSE5ermcQyBxGX",
        starCount: 10,
        teamId: "AD",
        username: "Jody"
      },
    ],
    starCount: 129,
    teamId: "AD"
  },
  {
    players: [
      {
        _id: "k3ezgBDwDZYj6wAZR",
        starCount: 40,
        teamId: "DM",
        username: "Fergus"
      },
      {
        _id: "WuyWdjRJNsAxegAuK",
        starCount: 30,
        teamId: "DM",
        username: "Quinlan"
      },
    ],
    starCount: 70,
    teamId: "DM"
  },
  {
    players: [
      {
        _id: "47YAyGoFTQ7wkztmj",
        starCount: 55,
        teamId: "UX",
        username: "Scott"
      },
      {
        _id: "iyoAjEoW7SApwjAaP",
        starCount: 12,
        teamId: "UX",
        username: "ScottFace"
      },
    ],
    starCount: 67,
    teamId: "UX"
  },
  {
    players: [
      {
        _id: "EGuZWhfieYvReZwCi",
        starCount: 25,
        teamId: "WD",
        username: "Sharifa"
      },
      {
        _id: "BaPqqbzf6uvTdfwAE",
        starCount: 0,
        teamId: "WD",
        username: "Kevin"
      },
    ],
    starCount: 25,
    teamId: "WD"
  },
  {
    players: [
      {
        _id: "d4NdHy3Nxqku7MN7S",
        starCount: 15,
        teamId: "UI",
        username: "Amir"
      },
    ],
    starCount: 15,
    teamId: "UI"
  },
]


if (Meteor.isClient) {
  describe('reusable functions - teamStarsFinder', function () {
    it('should take an array of objects, with a particular layout, as its input, do some work on the array, and compare its output with a manually entered array of objects, checking if they\'re deeply equal', function () {
      const teamStarsFinderOutput = teamStarsFinder(teams)
      chai.assert.deepEqual(teamStarsFinderOutput, lookingFor)
    })
    it('should not be deeply equal to a manually entered array that matches with teamStarsFinder\'s output but also has an additional object pushed onto the end of the array', function () {
      const teamStarsFinderOutput = teamStarsFinder(teams)
      let modifiedLookingFor = lookingFor
      modifiedLookingFor.push({teamId: 'face', starCount: 666})
      chai.assert.notDeepEqual(teamStarsFinderOutput, modifiedLookingFor)
    })
  })
}