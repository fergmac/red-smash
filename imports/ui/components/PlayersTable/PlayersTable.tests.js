import PlayersTable from './'
import React from 'react'
import { Meteor } from 'meteor/meteor'
import { shallow } from 'enzyme'
import { chai } from 'meteor/practicalmeteor:chai'
import { sinon } from 'meteor/practicalmeteor:sinon'

const tableData = [
  {
    id: "MRz6MwYfErHo9Aamx",
    starCount: 35,
    teamId: "AD",
    username: "Tracey",
  },
  {
    id: "47YAyGoFTQ7wkztmj",
    starCount: 55,
    teamId: "UX",
    username: "Scott",
  },
  {
    id: "k3ezgBDwDZYj6wAZR",
    starCount: 40,
    teamId: "DM",
    username: "Fergus",
  },
]

if (Meteor.isClient) {
  describe('component - PlayersTable', () => {
    describe('render', () => {
      it('should render', () => {
        const item = shallow(<PlayersTable players={tableData} />);
        chai.assert(item.hasClass('table-striped'));
      });

      it('should render the data passed into the table via props', () => {
        const item = shallow(<PlayersTable players={tableData} />);
        chai.assert(item.html().search('2nd') >= 0);
      });
    });

    describe('logic', () => {
      it('should call the numberSuffixer function', () => {
        numberSuffixer = sinon.spy()
        shallow(<PlayersTable players={tableData} />)
        expect(numberSuffixer.called)
      });
    });
  });
}