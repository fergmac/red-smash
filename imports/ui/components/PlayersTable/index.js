import React, { Component, PropTypes } from 'react';
import { Meteor } from 'meteor/meteor';
import { sortByKey, numberSuffixer } from '../../../functions'

export default class PlayersTable extends Component {
  render() {
    return (
      <table className="table table-striped">
        <thead>
          <tr className="face">
            <th className="text-left">Username</th>
            <th className="text-left">Team</th>
            <th className="text-left">Stars</th>
            <th className="text-left">Overall Rank</th>
          </tr>
        </thead>
        <tbody>
          {this.props.players.sort(sortByKey('starCount')).map((playerItem, index) => {
            return (
              <tr key={index} >
                <td className="text-left">{playerItem.username}</td>
                <td className="text-left">{playerItem.teamId}</td>
                <td className="text-left">{playerItem.starCount}</td>
                <td className="text-left">{numberSuffixer(index + 1)}</td>
              </tr>)
          })}
        </tbody>
      </table>
    );
  }
}

PlayersTable.propTypes = {
  players: PropTypes.array.isRequired,
};