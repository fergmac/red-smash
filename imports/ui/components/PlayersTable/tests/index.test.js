import React from 'react'
import PlayersTable from '../'
import renderer from 'react-test-renderer'
import { sortByKey, numberSuffixer } from '../../../../functions'

test('renders the PlayersTable component (a table, populated by props!)', () => {
  const players = [
    {
      username: 'Thomas',
      teamId: 'UX',
      starCount: 12
    },
    {
      username: 'Mike',
      teamId: 'UX',
      starCount: 8
    },
    {
      username: 'Zoe',
      teamId: 'AD',
      starCount: 14
    },
    {
      username: 'Barry',
      teamId: 'WD',
      starCount: 3
    },
    {
      username: 'Chloe',
      teamId: 'DM',
      starCount: 11
    },
    {
      username: 'Roberta',
      teamId: 'AD',
      starCount: 17
    },
    {
      username: 'Clarence',
      teamId: 'WD',
      starCount: 7
    },
    {
      username: 'Wilfred',
      teamId: 'UI',
      starCount: 15
    },
    {
      username: 'Audrey',
      teamId: 'WD',
      starCount: 4
    },
    {
      username: 'Donatello',
      teamId: 'UI',
      starCount: 9
    },
    {
      username: 'Kyle',
      teamId: 'DM',
      starCount: 19
    },
  ]

  const tree = renderer.create(
    <PlayersTable players={players} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
