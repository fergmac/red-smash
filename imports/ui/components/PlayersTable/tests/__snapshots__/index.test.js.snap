exports[`test renders the PlayersTable component (a table, populated by props!) 1`] = `
<table
  className="table table-striped">
  <thead>
    <tr>
      <th
        className="text-left">
        Username
      </th>
      <th
        className="text-left">
        Team
      </th>
      <th
        className="text-left">
        Stars
      </th>
      <th
        className="text-left">
        Overall Rank
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        className="text-left">
        Kyle
      </td>
      <td
        className="text-left">
        DM
      </td>
      <td
        className="text-left">
        19
      </td>
      <td
        className="text-left">
        1st
      </td>
    </tr>
    <tr>
      <td
        className="text-left">
        Roberta
      </td>
      <td
        className="text-left">
        AD
      </td>
      <td
        className="text-left">
        17
      </td>
      <td
        className="text-left">
        2nd
      </td>
    </tr>
    <tr>
      <td
        className="text-left">
        Wilfred
      </td>
      <td
        className="text-left">
        UI
      </td>
      <td
        className="text-left">
        15
      </td>
      <td
        className="text-left">
        3rd
      </td>
    </tr>
    <tr>
      <td
        className="text-left">
        Zoe
      </td>
      <td
        className="text-left">
        AD
      </td>
      <td
        className="text-left">
        14
      </td>
      <td
        className="text-left">
        4th
      </td>
    </tr>
    <tr>
      <td
        className="text-left">
        Thomas
      </td>
      <td
        className="text-left">
        UX
      </td>
      <td
        className="text-left">
        12
      </td>
      <td
        className="text-left">
        5th
      </td>
    </tr>
    <tr>
      <td
        className="text-left">
        Chloe
      </td>
      <td
        className="text-left">
        DM
      </td>
      <td
        className="text-left">
        11
      </td>
      <td
        className="text-left">
        6th
      </td>
    </tr>
    <tr>
      <td
        className="text-left">
        Donatello
      </td>
      <td
        className="text-left">
        UI
      </td>
      <td
        className="text-left">
        9
      </td>
      <td
        className="text-left">
        7th
      </td>
    </tr>
    <tr>
      <td
        className="text-left">
        Mike
      </td>
      <td
        className="text-left">
        UX
      </td>
      <td
        className="text-left">
        8
      </td>
      <td
        className="text-left">
        8th
      </td>
    </tr>
    <tr>
      <td
        className="text-left">
        Clarence
      </td>
      <td
        className="text-left">
        WD
      </td>
      <td
        className="text-left">
        7
      </td>
      <td
        className="text-left">
        9th
      </td>
    </tr>
    <tr>
      <td
        className="text-left">
        Audrey
      </td>
      <td
        className="text-left">
        WD
      </td>
      <td
        className="text-left">
        4
      </td>
      <td
        className="text-left">
        10th
      </td>
    </tr>
    <tr>
      <td
        className="text-left">
        Barry
      </td>
      <td
        className="text-left">
        WD
      </td>
      <td
        className="text-left">
        3
      </td>
      <td
        className="text-left">
        11th
      </td>
    </tr>
  </tbody>
</table>
`;
